File: changen.py
import json
import re

def format_iching_json(input_file, output_file):
    with open(input_file, 'r', encoding='utf-8') as f:
        content = f.read()

    # 移除开头和结尾的花括号，并分割每个卦象
    content = content.strip().strip('{}')
    hexagrams = re.split(r'},\s*{', content)

    formatted_data = {}
    for hexagram in hexagrams:
        try:
            # 为每个卦象添加花括号以创建有效的 JSON
            hexagram_json = json.loads('{' + hexagram + '}')
            formatted_data.update(hexagram_json)
        except json.JSONDecodeError as e:
            print(f"Error parsing JSON: {e}")
            print(f"Problematic string: {hexagram}")
            continue

    # 将格式化的数据写入新文件
    with open(output_file, 'w', encoding='utf-8') as f:
        json.dump(formatted_data, f, ensure_ascii=False, indent=4)

    print(f"格式化完成。输出文件：{output_file}")

# 使用脚本
input_file = 'standardized copy.txt'  # 您的输入文件名
output_file = 'formatted_iching_en.json'  # 输出文件名

format_iching_json(input_file, output_file)

File: changen.js
const fs = require('fs');

// Step 1: Read the file
const filePath = 'standardized_fr.txt';
const fileContent = fs.readFileSync(filePath, 'utf-8');

// Step 2: Split the content by lines
const lines = fileContent.split('\n');
console.log('Data split into lines:', lines);

// Step 3: Initialize an empty object to store the formatted data
let formattedData = {};

// Step 4: Loop through the lines and try to find and parse valid JSON blocks
let currentHexagram = null;
let currentKey = null;
let currentObject = null;

lines.forEach((line, index) => {
    // Skip empty lines
    if (line.trim() === '') {
        return;
    }

    // Step 4.1: Replace single quotes with double quotes to fix JSON formatting
    line = line.replace(/'/g, '"');

    // Check for the start of a new hexagram object
    if (line.includes('{')) {
        currentObject = '{'; // Start of new object
    }

    // Append lines that are part of the current object
    if (currentObject) {
        currentObject += line;

        // Check for end of the current hexagram object
        if (line.includes('}')) {
            try {
                const hexagram = JSON.parse(currentObject); // Try to parse the JSON
                const hexagramKey = Object.keys(hexagram)[0]; // Get the hexagram key (e.g., "15")

                // Add the parsed hexagram to the formatted data
                formattedData[hexagramKey] = hexagram[hexagramKey];
                console.log(`Parsed hexagram ${hexagramKey}:`, formattedData[hexagramKey]);

            } catch (error) {
                console.error(`Error parsing JSON at line ${index}:`, error);
            }

            currentObject = null; // Reset for the next hexagram
        }
    }
});

// Step 5: Write the formatted data to a new file
const outputFilePath = 'formatted_data.json';
fs.writeFileSync(outputFilePath, JSON.stringify(formattedData, null, 2), 'utf-8');

console.log('Data successfully saved as', outputFilePath);
